File: encompass-gc-optimized-collections/ts/gc_optimized_set.ts
/** @luaIterator */
export interface GCOptimizedSetIterable<T> extends Iterable<T> {}

export class GCOptimizedSet<TValue> {
    public static readonly Empty = new GCOptimizedSet<any>();

    get size(): number {
        return this.items.size;
    }

    private items: Set<TValue>; // Key type is actually TValue

    constructor(...items: TValue[]) {
        this.items = new Set<TValue>([...items]);
    }

    public entries(): GCOptimizedSetIterable<TValue> {
        return this.items;
    }

    public add(value: TValue): GCOptimizedSet<TValue> {
        this.items.add(value);
        return this;
    }

    public clear(): void {
        for (const k of this.items) {
            this.delete(k);
        }
    }

    public delete(value: TValue): boolean {
        return this.items.delete(value);
    }

    public has(value: TValue): boolean {
        return this.items.has(value);
    }
}

File: encompass-gc-optimized-collections/lua/gc_optimized_set.lua
local exports = {}
local GCOptimizedSet = {}
GCOptimizedSet.__index = GCOptimizedSet
GCOptimizedSet.prototype = {}
GCOptimizedSet.prototype.__index = GCOptimizedSet.prototype
GCOptimizedSet.prototype.constructor = GCOptimizedSet

GCOptimizedSet.new = function(...)
    local self = setmetatable({}, GCOptimizedSet.prototype)
    self:__constructor(...)
    return self
end

GCOptimizedSet.prototype.__constructor = function(self, ...)
    self.items = {}
    self.size = 0

    for _, item in ipairs({...}) do
        self.items[item] = true
        self.size = self.size + 1
    end
end

GCOptimizedSet.prototype.entries = function(self)
    return pairs(self.items)
end

GCOptimizedSet.prototype.add = function(self, item)
    if not self:has(item) then
        self.size = self.size + 1
    end
    self.items[item] = item
    return self
end

GCOptimizedSet.prototype.clear = function(self)
    for k in pairs(self.items) do
        self.items[k] = nil
    end
    self.size = 0
end

GCOptimizedSet.prototype.delete = function(self, item)
    if self:has(item) then
        self.items[item] = nil
        self.size = self.size - 1
    end
end

GCOptimizedSet.prototype.has = function(self, item)
    return self.items[item] ~= nil
end

GCOptimizedSet.Empty = GCOptimizedSet.new();

exports.GCOptimizedSet = GCOptimizedSet
return exports
